### PR TITLE
[maint_v3.x] V8 engine: X3D JavaScript calling Java

### PR DESCRIPTION
**Pull request by mwitchwilliams**
Enables a project for Tom to call Java methods from X3D &lt;SCRIPT&gt; node in JavaScript.
Tested with Tom's demo (though a retest would be good) and the X3D JavaScript demo's in the Test repo at https://github.com/gearvrf/GearVRf-Tests/tree/master/x3d/Scripting_Per_Frame and a subset of https://github.com/gearvrf/GearVRf-Tests/tree/master/x3d/Scripting

This Pull Request supersedes  Pull Request https://github.com/Samsung/GearVRf/pull/1596.  We will instead add support for v8 JavaScript engine for all GearVR in v4.x (master) instead.

GearVRF DCO signed off by: Mitch Williams
m1.williams@partner.samsung.com

### CODE REVIEW

**Comment by liaxim** [comment]
Is this still work in progress?

**Comment by mwitchwilliams** [comment]
This is Ready for Review.  It added the required functionality for maint_3.x release.  The portion that was still a work-in-progress will be part of v4.x instead.

File: GVRf/Framework/framework/src/main/java/org/gearvrf/script/GVRScriptManager.java [hunk 1]
```diff
@@ -217,19 +243,17 @@ public GVRScriptFile getScriptFile(IScriptable target) {
...
-        }
+        //final Object[] getV8Engine = new Object[0];
 
         GVRScriptFile script = null;
-        if (language.equals(LANG_LUA)) {
-            script = new GVRLuaScriptFile(mGvrContext, resource.getStream());
-        } else if (language.equals(LANG_JAVASCRIPT)) {
-            script = new GVRJavascriptScriptFile(mGvrContext, resource.getStream());
+        try {
+            InputStream inputStream = resource.getStream();
```
Comment (liaxim) [inline GVRf/Framework/framework/src/main/java/org/gearvrf/script/GVRScriptManager.java]: Precautionary you should use a try-with-resources statement. See https://docs.oracle.com/javase/tutorial/essential/exceptions/tryResourceClose.html

File: GVRf/Framework/framework/src/main/java/org/gearvrf/script/javascript/GVRJavascriptV8File.java [hunk 2]
```diff
@@ -40,14 +40,30 @@
     static protected ScriptEngine mEngine = null;
     protected Bindings bindings = null;
     protected Invocable invocable = null;
-    protected Bindings inputBindings = null;
-    Map inputVars = null;
+    static protected Bindings inputBindings = null;
```
Comment (liaxim) [inline GVRf/Framework/framework/src/main/java/org/gearvrf/script/javascript/GVRJavascriptV8File.java]: Why is inputBindings static and can it not be?

**Comment by liaxim** [comment]
disabled-gvr-javascript can quickly be patched not to use the gaze cursor controller so the javascript part can be exercised. It throws NPE for me:

```
11-20 09:56:58.083 26371 26371 E AndroidRuntime: FATAL EXCEPTION: main
11-20 09:56:58.083 26371 26371 E AndroidRuntime: Process: org.gearvrf.sample.gvrjavascript, PID: 26371
11-20 09:56:58.083 26371 26371 E AndroidRuntime: java.lang.RuntimeException: Unable to start activity ComponentInfo{org.gearvrf.sample.gvrjavascript/org.gearvrf.sample.gvrjavascript.GearVRJavascriptActivity}: java.lang.NullPointerException: Attempt to invoke virtual method 'boolean org.gearvrf.script.GVRScriptFile.invokeFunction(java.lang.String, java.lang.Object[])' on a null object reference
11-20 09:56:58.083 26371 26371 E AndroidRuntime: 	at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2984)
11-20 09:56:58.083 26371 26371 E AndroidRuntime: 	at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:3045)
11-20 09:56:58.083 26371 26371 E AndroidRuntime: 	at android.app.ActivityThread.-wrap14(ActivityThread.java)
11-20 09:56:58.083 26371 26371 E AndroidRuntime: 	at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1642)
11-20 09:56:58.083 26371 26371 E AndroidRuntime: 	at android.os.Handler.dispatchMessage(Handler.java:102)
11-20 09:56:58.083 26371 26371 E AndroidRuntime: 	at android.os.Looper.loop(Looper.java:154)
11-20 09:56:58.083 26371 26371 E AndroidRuntime: 	at android.app.ActivityThread.main(ActivityThread.java:6776)
11-20 09:56:58.083 26371 26371 E AndroidRuntime: 	at java.lang.reflect.Method.invoke(Native Method)
11-20 09:56:58.083 26371 26371 E AndroidRuntime: 	at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:1520)
11-20 09:56:58.083 26371 26371 E AndroidRuntime: 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1410)
11-20 09:56:58.083 26371 26371 E AndroidRuntime: Caused by: java.lang.NullPointerException: Attempt to invoke virtual method 'boolean org.gearvrf.script.GVRScriptFile.invokeFunction(java.lang.String, java.lang.Object[])' on a null object reference
11-20 09:56:58.083 26371 26371 E AndroidRuntime: 	at org.gearvrf.script.GVRScriptManager.attachScriptFile(GVRScriptManager.java:213)
11-20 09:56:58.083 26371 26371 E AndroidRuntime: 	at org.gearvrf.script.GVRScriptManager.bindHelper(GVRScriptManager.java:393)
11-20 09:56:58.083 26371 26371 E AndroidRuntime: 	at org.gearvrf.script.GVRScriptManager.bindScriptBundle(GVRScriptManager.java:293)
11-20 09:56:58.083 26371 26371 E AndroidRuntime: 	at org.gearvrf.sample.gvrjavascript.GearVRJavascriptActivity.onCreate(GearVRJavascriptActivity.java:78)
11-20 09:56:58.083 26371 26371 E AndroidRuntime: 	at android.app.Activity.performCreate(Activity.java:6955)
11-20 09:56:58.083 26371 26371 E AndroidRuntime: 	at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1126)
11-20 09:56:58.083 26371 26371 E AndroidRuntime: 	at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2927)
11-20 09:56:58.083 26371 26371 E AndroidRuntime: 	... 9 more
```

**Comment by mwitchwilliams** [comment]
Ready for Review. Fixed issues in Tom's demo.
Developers using JavaScript to build GearVR scenes, such as in gvr-JavaScript demo above, should use master v4.x release.

**Comment by liaxim** [comment]
Ok, planning to merge on Tuesday but it seems to me nothing but x3d now works with javascript.

Tried gvr-remotescripting, after the ``importPackage(org.gearvrf);`` statement the app crashes with:
```
12-04 07:14:36.954 18079 18079 F DEBUG   : *** *** *** *** *** *** *** *** *** *** *** *** *** *** *** ***
12-04 07:14:36.955 18079 18079 F DEBUG   : Build fingerprint: 'samsung/dreamqltesq/dreamqltesq:7.0/NRD90M/G950USQU1AQFJ:eng/test-keys'
12-04 07:14:36.955 18079 18079 F DEBUG   : Revision: '12'
12-04 07:14:36.955 18079 18079 F DEBUG   : ABI: 'arm'
12-04 07:14:36.956 18079 18079 F DEBUG   : pid: 18016, tid: 18073, name: pool-1-thread-1  >>> org.gearvrf.sample.remote_scripting <<<
12-04 07:14:36.956 18079 18079 F DEBUG   : signal 11 (SIGSEGV), code 1 (SEGV_MAPERR), fault addr 0xe60
12-04 07:14:36.956 18079 18079 F DEBUG   :     r0 c095846c  r1 00000000  r2 00000000  r3 00000000
12-04 07:14:36.956 18079 18079 F DEBUG   :     r4 c095846c  r5 00000001  r6 00000000  r7 c0958448
12-04 07:14:36.956 18079 18079 F DEBUG   :     r8 c09585b8  r9 c8d96900  sl 00000000  fp c095853c
12-04 07:14:36.957 18079 18079 F DEBUG   :     ip cdbeb2ec  sp c0958418  lr cd12e403  pc cd501310  cpsr 400e0010
12-04 07:14:36.968 18079 18079 F DEBUG   : 
12-04 07:14:36.968 18079 18079 F DEBUG   : backtrace:
12-04 07:14:36.968 18079 18079 F DEBUG   :     #00 pc 0085b310  /data/app/org.gearvrf.sample.remote_scripting-2/lib/arm/libjav8.so (_ZN2v86Locker10InitializeEPNS_7IsolateE+48)
12-04 07:14:36.968 18079 18079 F DEBUG   :     #01 pc 004883ff  /data/app/org.gearvrf.sample.remote_scripting-2/lib/arm/libjav8.so (_ZN5V8EnvC2EP7_JNIEnv+62)
12-04 07:14:36.968 18079 18079 F DEBUG   :     #02 pc 004882d7  /data/app/org.gearvrf.sample.remote_scripting-2/lib/arm/libjav8.so (Java_lu_flier_script_V8CompiledScript_internalCompile+30)
12-04 07:14:36.969 18079 18079 F DEBUG   :     #03 pc 004dd355  /data/app/org.gearvrf.sample.remote_scripting-2/oat/arm/base.odex (offset 0x482000)
```

**Comment by mwitchwilliams** [comment]
Tom's work which uses X3D with JavaScript must be run on the same thread. For the moment, we are running the v8 engine on GLthread. In v4.x, I expect to code a new JSthread for JavaScript.
The recommendation is for those currently using JavaScript but not in X3D to go to master, v4.x

I can code this so both work in maint_v3.x.  This would involved having duplicates of several methods - one for X3D running JS on v8 using the GLthread, and the other for JavaScript applications still on the Rhino engine such as gvr-remotescripting and gvr-javascript. But again, these JavaScript applications currently work on v4.x so it seemed a bit of effort for an older release (though it would help in developing the v4.x version).

I had begun the work on the JSthread and will return to it, but it will a big effort to code.

**Comment by liaxim** [comment]
I am not sure how can we make a maintenance 3.x release in this state. Hopefully we will not have to.

**Comment by mwitchwilliams** [comment]
I think you have a valid concern.
Not only does this PR break some gvr demo's in maint-_v3.x, but I assume this will eventually be merged into master v4.x

Let me propose the 2 fixes:

1. A few new methods in GVRScriptManager.java would have to be duplicated.
2. Implement the JSthread

#1 is the simplest. I would write a new constructor, initializeEngines(), addGlobalBindings(), and addVariable(). Tom would need to modify his demo to call a different addVariable().

#2 The new JSthread so far has been a tedious implementation. X3D JavaScript threads have can have an initialize() method and other init methods performed after parsing the X3D file, but before rendering the first frame.  We will need to make this work eventually. I anticipate it will take some rework and extensive testing to make this modification.

Perhaps we can visit this issue with Tom.

**Comment by liaxim** [comment]
#1 seems the most appropriate for the maint branch.

**Comment by liaxim** [comment]
@mwitchwilliams My understanding is that you are looking into improving this change so JS works outside of X3D. Correct?

**Comment by mwitchwilliams** [comment]
At this very moment, I am assisting Eric's team. I believe Tom cherry-picked this code for his needs.

Perhaps, we can discuss with Tom if this PR should work with all JS demo's in maint_v3.x release and thus can be merged into v4.x; or should we just begin with v4.x as a first of a multi-step process migrating all GearVR from Rhino engine to v8 JS engine on a new 'JSthread'.

Another recent issue is the v8 engine, which needs to be a singleton, has to shut down upon exiting GearVR apps.

I believe the fix will be relatively simple to have this code work with all JS demo's in maint_v3.x, and then can be merged into v4.x, before I make the major changes in v4.x

**Comment by liaxim** [comment]
I don't know why v8 needs to be a singleton but it sounds like a bad idea to me.

**Comment by mwitchwilliams** [comment]
Tested with gvr-javascript, gvr-remote-scripting, gvr-x3d-360video and various demos from the GearVr-Test repo in the X3D folder that included JavaScript, and a few other of my test files.

Tom will have modify X3D360Video.java file, moving 

```
        GVRScriptManager scriptManager = gvrContext.getScriptManager();
        scriptManager.addVariable("player", new PlayerInterface(gvrContext));
```
below the try-catch statement that loads the X3D file.  The reason for this change is that GearVRf initialization calls GVRViewManager.java which constructs GVRManager before we know we have an X3D file. When we launch the X3D file, We now tell GVRScriptManager to use v8 engine.  This of course is a temporary situation and will be resolved when we implement v8 for all GearVR, presumably in v4.x release..

File: GVRf/Framework/framework/src/main/java/org/gearvrf/script/GVRScriptManager.java [hunk 3]
```diff
@@ -170,10 +218,37 @@ public ScriptEngine getEngine(String language) {
...
+        }
+        else {
+            final String varNameFinal = varName;
+            final Object valueFinal = value;
+            mGvrContext.runOnGlThread(new Runnable() {
+                @Override
+                public void run() {
+                    ScriptEngine engine = getEngine(LANG_JAVASCRIPT);
+                    // need to add the package where this variable is being added from
+                    int lastPeriod = valueFinal.toString().lastIndexOf('.');
```
Comment (liaxim) [inline GVRf/Framework/framework/src/main/java/org/gearvrf/script/GVRScriptManager.java]: The return value of lastIndexOf should be checked and dealt with accordingly

File: GVRf/Framework/framework/src/main/java/org/gearvrf/script/javascript/GVRJavascriptV8File.java [hunk 4]
```diff
@@ -41,20 +41,29 @@
...
 
-    public GVRJavascriptV8File(GVRContext gvrContext, String scriptText) {
+    public void setExternalImportStatement(String externalImportStatement) {
+        this.externalImportStatement = externalImportStatement;
+    }
 
-        mScriptText = scriptText;
-        mGvrContext = gvrContext;
 
-        GVRContext.addResetOnRestartHandler(new Runnable() {
```
Comment (liaxim) [inline GVRf/Framework/framework/src/main/java/org/gearvrf/script/javascript/GVRJavascriptV8File.java]: You need to add back the addResetOnRestartHandler for the static mEngine

**Comment by liaxim** [comment]
Javascript-using apps work now.

**Comment by liaxim** [comment]
Looks good to me. It can be merged.

**Comment by liaxim** [comment]
@mwitchwilliams What is that status of this? Is it waiting on something or can we double-check and merge?

**Comment by mwitchwilliams** [comment]
This is ready to merge. It is not waiting on anything.  However, I will do a recheck since I am coding X3D's Movie Texture and Tom's demo provides some guidance in implementing similar features.

**Comment by mwitchwilliams** [comment]
Rested with Tom's gvr-x3d-360video annd gvr-javascript and they worked fine. So...
**LGTM (Looks Good To Me).**

**Comment by liaxim** [comment]
Planning to merge on Thursday

**Pull request merged**